HPCC-10618 Add support for MongoDB

delvecja opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
Added MongoDB Plugin

Fixed Deserialization Error and added option for using connection options when connecting to Cluster.

Force users to provide a username and password when connecting to MongoDB Atlas, and forbid it when connecting to a local mongod instance.
Refactor deserializeEJSON method. Add helper method for capturing EJSON as standard JSON.

Add simple ecl file for regression testing.

Add else statement for unsupported types.
Change user in mongodb-simple.ecl to hpccuser.
Fix formatting issue.

Add comment detailing how to load sample datasets onto a MongoDB Atlas cluster.

Fix wording of comment.

Add key for regression test.
Add status to end of mongodb-simple.ecl.

Fix careless mistake causing error when compiling.

Add additional conditions to ensure return documents are always the same.

Regression tests return exactly the same documents each time.
Correct error in key file, and ensure all entries are correct.

Address minor changes

Force users to provide id.

Signed-off-by: jackdelv <jackdelv39@gmail.com>

Plugin handles leading ampersand and question mark when specifying connectionOptions.
Clarify MongoDB operators in README.md and link to manual showing all available operators.

Fix style errors.

Fix additional style errors.

Add const keyword for string in insertPair.
Fix getRowResult to return pointer to next row rather than using getLink().

Fix UTF8 translation from MongoDB result to ECL records.

Define end as const char * rather than auto.
Add checks to all loops to ensure pointer doesn't run off end of string.

nextRow() now returns null pointer rather than initializing result just to return the nullptr.

Added check for opening parenthesis and safeguard in while loop so it never runs off the end of the string.

Handle connection options properly.
Check for opening parenthesis properly.

Remove redundant chars parameter.
Minor: change variable name to be more suitable.

Fix bindDataParam string construction

Handle string and utf conversions more carefully.

Remove unused constructor in Pathtrackeer class.

Added default initializers to some members in MongoDBRecordBinder.

minor: add new line at end of file.

Add Copyright header and smoketest flags.

richardkchapman commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+                depth++;
+                lastBrkt = start; // Keep track of last bracket in case we need to backtrack
+            }
+            if (*start == '$' && *(start-1) == '\"') // Look for '$' to mark the start of a datatype
```
This may access *(row-1) which is not guaranteed to be readable, and may core.

richardkchapman commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+            if (*start == '$' && *(start-1) == '\"') // Look for '$' to mark the start of a datatype
+            { 
+                auto end = start;
+                while (*end != '\"') end++;
```
Will this read off the end of the string if no " is found?

delvecja replied:
Technically yes, if no " is found it reads off the end, but because we are only reading MongoDB EJSON objects the datatype names are always surrounded by ". An example of the objects this function processes is:
`{"$date":{"$numberLong":"-1577923200000"}}`

rpastrana commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+                } 
+                else if (key == "$timestamp") 
+                {
+                    while (*end != '}') end++; // Skip over timestamp
```
are there other chars which if encountered, should force a break out of the loop? should we ever iterate past string terminator, or len of string-1?

delvecja replied:
For this case we are looking at an object of this form:`{"someTimestamp": {"$timestamp":{"t":1565545664,"i":1}}`. Since it is not trivial to convert this to an ECL data type I simply skip over it and replace it with an empty string. I believe for this case we will only want to break the loop when a '}' is encountered and we need to do that twice for timestamps. There shouldn't be a reason where we need to iterate past the string terminator or len of string-1.

rpastrana commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+        int64_t mongoVal = val;
+        std::string terminatedString = name;
+
+        query->build()->append(kvp(terminatedString, bsoncxx::types::b_int64{mongoVal}));
```
trivial, but likely a little cleaner to create the std::string inline as was done in other functions:
query->build()->append(kvp(std::string(name)

delvecja replied:
Changed to inline.

rpastrana reviewed (changes requested):
Overall looks good, please take a look at my comments/questions.

delvecja commented on plugins/mongodb/CMakeLists.txt:
```diff
@@ -0,0 +1,94 @@
+IF(MONGODBEMBED)
+    ADD_PLUGIN(mongodbembed)
+    if(MAKE_MONGODBEMBED)
+        find_package(bsoncxx CONFIG REQUIRED)
```
I am not very familiar with the vcpkg stuff. As far as I can tell the mongo-cxx-driver depends on libbson which is what bsoncxx is built from.

delvecja commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,693 @@
+    static void typeError(const char *expected, const char * fieldname)
+    {
+        VStringBuffer msg("MongoDB: type mismatch - %s expected", expected);
+        if (fieldname && *fieldname)
```
The if statement is now 
        if (!isEmptyString(fieldname))

delvecja commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,693 @@
+            case Deserialization_SUCCESS:
+                break;
+            case Deserialization_BAD_TYPE:
+                failx("Deserialization error (%s): value cannot be const", targetype);
```
That should be correct although I took this method directly from the couchbase plugin, so I'm not too familiar with the DeserializationResult enums.

delvecja commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,247 @@
+
+## Documentation
+
+[Doxygen](https://www.doxygen.nl/index.html) can be used to create nice HTML documentation for the code. Call/caller graphs are also generated for functions if you have [dot](https://www.graphviz.org/download/) installed and available on your path.
```
Developers can supply their own Doxyfile, but there were a few changes I made to show an examples page, and setting the homepage to the README.md file. The graphviz was meant to be a supplement as it only adds some extra figures and call/caller graphs. I think the graphviz graphs were helpful and if there is more I can do with it I would be happy to look into graphviz more.

rpastrana replied:
@JamesDeFabia ??

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+using bsoncxx::builder::stream::array_context;
+using bsoncxx::builder::basic::kvp;
+
+static const char *MODULE_NAME = "mongodb";
```
I believe the MODULE_NAME is what is used in the IMPORT statement. (I might be wrong) I think having `IMPORT MongoDB` looks better than `IMPORT MongoDBEmbed`, but if you think the latter is more descriptive than that is the better one.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+        {
+            failx("Duplicate Key: %s",e.what());
+        }
+        failx("Exception: %s", e.what());
```
I am actually not sure. It returns a char const* which could be empty. I have changed it to check for an empty string.

if (!isEmptyString(e.what))

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+                result += std::string(row, lastBrkt - row); // Add everything before we went into nested document
+                if (key == "$regularExpression") // Some data types are unsupported as they are not straightforward to deserialize
+                {
+                    UNSUPPORTED("Regular Expressions are currently Unsupported"); // TO DO handle unsupported types by not throwing an exception.
```
I was actually using the UNSUPPORTED method incorrectly. It already prefixes the output with "Unsupported Feature:". I have changed the UNSUPPORTED message to 
UNSUPPORTED("MongoDB Regular Expressions");

delvecja replied:
After making a change to the UNSUPPORTED method to prefix the error message with MODULE_NAME I changed the UNSUPPORTED message in deserializeEJSON to 
UNSUPPORTED("Regular Expressions");

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+using bsoncxx::builder::basic::kvp;
+
+static const char *MODULE_NAME = "mongodb";
+static const char *MODULE_DESCRIPTION = "MongoDB Embed Helper";
```
Changed L66-69 to static constexpr const char *

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+     */
+    extern void UNSUPPORTED(const char *feature)
+    {
+        throw MakeStringException(-1, "UNSUPPORTED feature: %s not supported in %s", feature, VERSION);
```
Changed to
throw MakeStringException(-1, "%s UNSUPPORTED feature: %s not supported in %s", MODULE_NAME, feature, VERSION);

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+        }
+
+        if (port != 0)
+            connectionString.appendf("mongodb://localhost:%i", port); // Connection to a local MongoDB instance
```
No, the Atlas clusters operate on specific ports that do not need to be set. The Atlas clusters are MongoDB's hosted clusters, but it is possible to have a local instance of MongoDB running on local host in which case you need to set the port for connecting to it. Changed to

`if (!isEmptyString(server.c_str()) && port !=0)
    failx("Specifying a port is not allowed when connecting to an Atlas Cluster."
else if (port != 0)`

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+    MongoDBEmbedFunctionContext::MongoDBEmbedFunctionContext(const IContextLogger &_logctx, const char *options, unsigned _flags)
+    : logctx(_logctx), m_NextRow(), m_nextParam(0), m_numParams(0), m_scriptFlags(_flags)
+    {
+        const char *server = "";
```
MongoDB does accept connection options such as those, but they need to be specified in the connection string which is supplied by the user. Since it has to be up to the user adding this information to the README.md file will be helpful.

rpastrana replied:
ok, let's def add that info to README.md, and the examples, but it looks like you're constructing a connection string on line 885 based off input options...

delvecja replied:
Yes that is an oversight on my part. I will add an additional option for the User to specify any number of connection options and add that to the connection string.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+        StringBuffer connectionString;
+        count = 0; // Initialize count as 0
+        // StringBuffer connectionOptions;
+        // unsigned int maxConnections = 0;
```
Removed commented lines as connection options are specified in the URI connection string.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+        bool useSSL = false;
+        StringBuffer connectionString;
+        count = 0; // Initialize count as 0
+        // StringBuffer connectionOptions;
```
Removed.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+     */
+    size32_t MongoDBEmbedFunctionContext::getTransformResult(ARowBuilder & rowBuilder)
+    {
+        // To Do
```
Changed TODOs to report unimplemented.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+     * @return true Result Rows has a length greater than 0.
+     * @return false If no result was returned and the result rows are empty.
+     */
+    bool MongoDBEmbedFunctionContext::getBooleanResult() 
```
MongoDB does not return scalar values it only returns JSON objects so there isn't necessarily a next scalar. The order of these documents and their fields is not guaranteed, so an ECL user trying to get a single scalar field from the document will not always get the same value. For these scalar return types I was able to find a few datatypes that could represent something useful about the MongoDB query rather than returning a random scalar value from the MongoDB document/JSON object.

rpastrana replied:
Fair enough.
So it sounds but it sounds like this operation should be UNSUPPORTED

delvecja replied:
Yeah that makes a lot of sense. The query throws an exception if it fails, so returning the boolean is redundant. I changed the other scalar types to unimplemented as well. For the insertMany function it returns a count of the inserted documents so I return that as a result row rather than a scalar now. I added two more standard record layouts to the ecllib file to avoid someone messing up on the name of the field that is returned.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2302 @@
+        while (*start != '[') start++;
+        auto builder = document{};
+
+        
```
Fixed.

delvecja commented:
@rpastrana Ok I believe I have responded to all of your comments and made the appropriate changes. I've replied to your comments with the changes I've made before I make another commit.

delvecja commented:
@richardkchapman I have responded to both of your concerns. Take a look at my comments and let me know what you think. Thanks!

rpastrana reviewed (changes requested):
A couple follow up comments.

rpastrana reviewed (changes requested):
@delvecja added a few comments

delvecja commented:
@rpastrana @richardkchapman Thank you for your patience it has been a busy couple of weeks.

I have fixed an error in the deserialization method and tested it on over 10000 MongoDB documents to ensure its stability. 

I have added the connection options to the list of options that are used to build the connection string. I have also updated the documentation to show the intricacies of it so there should be no confusion on the part of the ECL user. 

In the deserialization method I added conditions to make sure none of the open loops run past the end of the string.

rpastrana commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2287 @@
+                depth++;
+                lastBrkt = start; // Keep track of last bracket in case we need to backtrack
+            }
+            if (*start == '{' && *(start + 1) == ' ' && *(start + 2) == '\"' && *(start + 3) == '$') // Look for "{ \"$" to mark the start of a datatype
```
it feels like this block should be within the the if block starting in L161 (for readability/maintainability)

delvecja replied:
I agree. I will refactor that code to make it more readable because currently it is dense and not clear what the purpose behind it is.

rpastrana commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2287 @@
+        else if (port != 0)
+            connectionString.appendf("mongodb://localhost:%i/%s", port, connectionOptions); // Connection to a local MongoDB instance
+        else
+            connectionString.appendf("mongodb+srv://%s:%s@%s/?retryWrites=true&w=majority%s", user, password, server, connectionOptions); // Connection to MongoDB Atlas server
```
I think it's possible for user and pass to be empty, we should either require those values, or avoid the user:pass@ portion in that scenario.

delvecja replied:
Correct they can be empty when connecting to a local instance, but a username and password are required to connect to a MongoDB cluster. I will change it so that they are required for connecting remotely and not allowed when connecting locally.

rpastrana reviewed (changes requested):
@delvecja a couple of comments, but also, please add regression tests, see this commit for reference:
https://github.com/hpcc-systems/HPCC-Platform/commit/5e7502d1ea7249ddca3e94bd241b5f5b1ff32a27#diff-0082fb0e4587dae424df77beafe0df5067329d9b95af6703d7abc18d2b014689

delvecja commented:
@rpastrana I have addressed your two comments and moved on to the regression testing. I was wondering how the regression tests would connect to a MongoDB instance since either you need a user and an Atlas cluster for the regression testing or you need to set up a local mongod instance which is not trivial.

rpastrana commented:
@delvecja setting up the servers for the regression tests can be a challenge, but @AttilaVamos has done this for a few of the embedded ECL plugins. Can you work with him on best/simplest way to setup a running mongodb in our smoketest environment?

delvecja commented:
@rpastrana @AttilaVamos Of course! The easiest way to set it up would be to have a LexisNexis Account sign up for the free MongoDB Atlas tier. That is very straightforward and then all you would need to do is configure the user with a username and password and copy the address of the cluster. Once the MongoDB Atlas is set up it has to be loaded with the sample dataset that is used for the test queries which is also very easy. The difficult part of that would be ensuring that the username, password, and server address are not exposed in the commit. Other than that though I don't see any other difficulties as it would then be configured the same way as my school MongoDB account which I used for testing.

rpastrana commented:
@delvecja let's do something similar to the mysql plugin:
https://github.com/hpcc-systems/HPCC-Platform/blob/f4331e19103a0f14441334f31b3e5fc23627e9bf/testing/regress/ecl/mysqlembed.ecl#L25
we can default to localhost,  and some dumb/default user/pass.

When @AttilaVamos  sets up the server, we could adjust.

delvecja commented:
@rpastrana Ok, I will get it going. I should be able to get that finished by tomorrow. I will push the commit with your other two comments when the regression tests are done.

rpastrana commented:
Thanks @delvecja

rpastrana commented:
@delvecja before I forget, please don't squash your code review commits so it's easier to see the latest changes. Once completed we'll ask you to squash all commits. Thank.

rpastrana reviewed (changes requested):
@delvecja we're getting close, but I found a couple of issues.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -895,12 +909,29 @@ namespace mongodbembed
+            if (isEmptyString(user) && isEmptyString(password))
+                connectionString.appendf("mongodb://localhost:%i/%s", port, connectionOptions); // Connection to a local MongoDB instance
+            else
+                failx("Username and Password is not allowed when connecting to local mongod instance");
```
"mongod" is the name for the process in charge of handling client connections to a local server. I agree that it is strange that a user is not allowed and unfortunately the documentation for a local instance of MongoDB is much more sparse than for connecting to an Atlas server. From what I have seen the connection string does not contain any user information, but I will look into that further.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -895,12 +909,29 @@ namespace mongodbembed
+            if (!isEmptyString(user) && !isEmptyString(password))
+                connectionString.appendf("mongodb+srv://%s:%s@%s/?retryWrites=true&w=majority%s", user, password, server, connectionOptions); // Connection to MongoDB Atlas server
+            else 
+                failx("Username or Password not supplied. Use the user() or password() options in the EMBED declaration.");
```
Maybe because I wrote it I can't see where the error is unless it should be "Username and Password --is-- not supplied."

rpastrana replied:
Ignore.

rpastrana commented on testing/regress/ecl/mongodb-simple.ecl:
```diff
@@ -0,0 +1,96 @@
+
+SEQUENTIAL
+(
+    OUTPUT(findTransactionsByID(716662), NAMED('FindTransactionByID'));
```
ideally the output of these commands should be reproducible - is there sample/generated dataset(s) required for these commands to execute as expected?

delvecja replied:
Yes, there is an option in MongoDB atlas to load a "sample dataset" which includes multiple collections. I have loaded it on two different accounts and they were the same datasets, but there is nothing stopping MongoDB from changing the sample datasets that they include in the future.

rpastrana replied:
Fair enough. So are these commands expecting one of those datasets?
Regardless of the source of the data, we should ensure we document how to load the data these commands expect.

delvecja replied:
Yes, these commands are expecting to see the sample datasets from MongoDB, but the way that MongoDB operates is if you specify a collection that does not exist it will create it on the cluster. In order to load every dataset that the commands are expecting there is only a single step which I could create some documentation for. I believe it is trivial and MongoDB already offers documentation for it, but I could create some additional documentation that would make it even simpler since I have done it a few times.

rpastrana replied:
Got it!
Ok so, sample_airbnb.listingsAndReviews and  sample_airbnb.transactions are made available on the cluster automatically.
I think a quick comment here will go a long way.

delvecja replied:
Yes, definitely! I will add a comment so there is no confusion.

rpastrana reviewed (changes requested):
@delvecja a few more comments re: test suite.

delvecja commented on testing/regress/ecl/mongodb-simple.ecl:
```diff
@@ -0,0 +1,96 @@
+
+server := 'cluster0.uniqueserver.mongodb.net';
+
+username := 'delvecja';
```
Changed to hpccuser.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -130,6 +130,98 @@ namespace mongodbembed
+            depth--;
+            row = end;
+            start = end;
+        }
```
added failx("EJSON datatype error: %s is not supported in the current version.", key); in else block.

delvecja commented:
@rpastrana I added the key for the regression test, but admittedly I rushed through it because I need to get started on some work for my classes. I will go over mongodb-simple.xml and mongodb-simple.ecl in more depth when I can find the time, hopefully later tonight. I also still need to confirm the connection options for connecting to a local instance of MongoDB running mongod.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -130,6 +130,98 @@ namespace mongodbembed
+        result += std::string(row, lastBrkt - row); // Add everything before we went into nested document
+        if (key == "$regularExpression") // Some data types are unsupported as they are not straightforward to deserialize
+        {
+            UNSUPPORTED("Regular Expressions"); // TO DO handle unsupported types by not throwing an exception.
```
I cannot view that issue. "It may have been deleted or you don't have permission to view it." Is what I am seeing.

rpastrana replied:
No problem, it's not assigned to you (unless you want to take it on), we'll change its visibility to community soon.

delvecja replied:
I'll keep an eye on it then. Obviously anything involving this plugin I would be happy to help with in the future, and since I had such a positive experience with the platform I will definitely be looking for more ways to contribute.

richardkchapman replied:
I have made that JIRA ticket public

AttilaVamos commented:
Two comments:
- In GitHub Smoketest action we don't test (yet) those plugins which need 3rd party server like MySQL,, Couchbase, MongoDB, etc.
- I see lots of regression on CentOS 7 environment with all engines. Most of errors are timeout. I don't know yet why and they seems totally unrelated test cases to this PR:
```
 aggidx1.ecl ( version: multiPart=true ),
 aggidx3.ecl ( version: multiPart=true ),
 bug5236.ecl ( version: multiPart=true ),
 bug5236.ecl ( version: multiPart=true, useTranslation=true ),
 countindex.ecl ( version: multiPart=true ),
 countindex.ecl ( version: multiPart=true, useTranslation=true ),
and so on
```

delvecja commented:
@AttilaVamos My apologies Attilla, I erroneously pushed that last change without trying to compile. There shouldn't be any further compilation issues as I have ran my example test case and it succeeded.

rpastrana commented:
> Two comments:
> 
>     * In GitHub Smoketest action we don't test (yet) those plugins which need 3rd party server like MySQL,, Couchbase, MongoDB, etc.
> 
>     * I see lots of regression on CentOS 7 environment with all engines. Most of errors are timeout. I don't know yet why and they seems totally unrelated test cases to this PR:
> 
> 
> ```
>  aggidx1.ecl ( version: multiPart=true ),
>  aggidx3.ecl ( version: multiPart=true ),
>  bug5236.ecl ( version: multiPart=true ),
>  bug5236.ecl ( version: multiPart=true, useTranslation=true ),
>  countindex.ecl ( version: multiPart=true ),
>  countindex.ecl ( version: multiPart=true, useTranslation=true ),
> and so on
> ```

@AttilaVamos we recently discussed test involving 3rd party platforms (OBT test?)
Is there anything keeping us from testing this in OBT in the future?

AttilaVamos commented:
@delvecja Those errors are regression errors instead of compile time problems.

AttilaVamos commented:
@rpastrana As soon as this PR will be merged and I can have stable environment where OBT it can build the platform from 8.8.x and beyond versions, then we can test this plugins along with the others.

delvecja commented:
@rpastrana I went back through and corrected some of the mistakes and added more rigorous tests to ensure the same documents are getting returned each time.

rpastrana commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -222,7 +222,7 @@ namespace mongodbembed
         else
         {
-            failx("EJSON datatype error: %s is not supported in the current version.", key);
+            failx("EJSON datatype error: %s is not supported in the current version.", key.c_str());
```
minor, but I find it helpful to encapsulate the %s placeholder in case key is empty or very long

delvecja replied:
Do you mean something like this?
failx("EJSON datatype error: %s is not supported in the current version.", (!key.empty() || key.size() < 25) ? key.c_str() : "Unknown Datatype");

delvecja replied:
That should be &&, but is that max length good? Should I throw another error all together?

rpastrana replied:
I just meant something like this:
failx("EJSON datatype error: '%s' is not supported in the current version.", key.c_str());

delvecja replied:
Ok, much simpler.

rpastrana reviewed (commented):
@delvecja looks good. I left a couple of minor comments. once done let's sqash for clean merge.Thanks again

delvecja commented:
@rpastrana I got everything squashed. Passing it back to you.

delvecja commented:
@AttilaVamos The regression test for MongoDB that is failing is expected to fail because the server that it is trying to connect to does not exist. Should I add a separate key file for this error message until the MongoDB Atlas server gets set up?

rpastrana reviewed (commented):
@delvecja just a couple more minor issues

delvecja commented on testing/regress/ecl/mongodb-simple.ecl:
```diff
@@ -0,0 +1,92 @@
+ENDEMBED;
+
+// Returns a dataset. Finds the document where the beds and accomodates fields are higher than the argument values. Matches _id to ensure the same document gets returned.
+dataset(airbnbrec) findABBeds(INTEGER beds, INTEGER accommodates, STRING id='10006546') := EMBED(mongodb : user(username), password(password), server(server), database('sample_airbnb'),  collection('listingsAndReviews'))
```
Unfortunately, it cannot ignore the parameter if it is an empty string. It will create the document with the _id field having an empty string for its value which will match 0 zero documents. I could remove the _id parameter altogether, but then it will be possible for different documents to be returned.

delvecja replied:
I could also change the commands to use different fields to get it down to a single document without having to use _id.

rpastrana replied:
can we use a different version of find_one({beds: {$gte: $beds}, accommodates: {$gte: $accommodates}); without "_id" if id is empty?
if not, it's not a big deal, force the caller to provide the proper id like you were doing before.

delvecja replied:
I can do that, but I would have to create an additional findFeebyBeds function that has a different embed statement and use the ECL IF statement to decide which of the two functions to call. The plugin is kind of "dumb" and it only allows a single line in the embedded script and uses that assumption to build the documents. If that is favorable I can get going on that.

rpastrana replied:
fair enough, I'd force the user to provide the ID value

delvecja replied:
Ok, I will make that change.

delvecja commented:
@rpastrana back to you Rodrigo. Also, I squashed it by mistake my apologies.

rpastrana commented:
Approved

ghalliday commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+        if (*start == '\"') 
+        {
+            end = ++start;
+            while (*end != '\"') end++;
```
Is the input guaranteed to be wel-formed, or could this loop past the end of a invalid string?

delvecja replied:
This is input from the embedded script, so there is a good chance that it will not be well-formed. For loops like this I can add something like while (*end && *end != '\"').

ghalliday commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+        if (!value || !*value) 
+        {
+            NullFieldProcessor p(field);
+            rtlStrToStrX(chars, result, p.resultChars, p.stringResult);
```
rtlUtf8ToStringX

delvecja replied:
Fixed

ghalliday commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+        if (!value || !*value) 
+        {
+            NullFieldProcessor p(field);
+            rtlStrToStrX(chars, result, p.resultChars, p.stringResult);
```
rtlUtf8ToUtf8X

delvecja replied:
Fixed

ghalliday commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+                UNSUPPORTED("IMPORT");
+                return nullptr;
+            } 
+            else return new MongoDBEmbedFunctionContext(ctx ? ctx->queryContextLogger() : queryDummyContextLogger(), options, flags);
```
style: statement on following line.

delvecja replied:
Fixed style errors.

ghalliday reviewed (changes requested):
Broadly the code looks fairly good.
* A few issues with understanding/processing ECL data types which are quite understandable.
* Some style issues.  Because this is an external contribution in an independent chunk of code they can be left as they are - I haven't commented on them all (mainly statements on the same line as ifs and while loops).
* A few concerns about handling invalid input.  It would be good to have some test cases that include processing blank embeds, invalid bracket/quote nesting and other code that could potentially cause problems.
* Not sure about the structure of the classes.  For instance, if there are multiple uses of mongo db to different servers within the same query (or multiple queries in roxie).

I think it would be worth addressing the errors cases and ECL type encoding issues, and create a separate jira to tackle shared uses of mongodb since that is likely to need some thought, and the plugin will still be useful as it stands.

delvecja commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,269 @@
+
+#### Connection Options 
+
+To specify connection options to the MongoDB cluster use the connectionOptions option in the embed statement. The format for the connection options is a question mark followed by anpersant separated options like so: ?<option>&<option>
```
Changed

delvecja commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,269 @@
+
+```
+connectionOptions('&ssl=true&connectTimeoutMS=1000') // For connecting to MongoDB Clusters
+connectionOptions('?ssl=true&connectTimeoutMS=1000') // For connecting to a local mongod instance
````
Plugin now adds leading '?' or '&'

delvecja commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,269 @@
+
+```
+dataset({STRING _id}) getCount(REAL salary) := EMBED(mongodb : user(user), password(pwd), server(server), database(dbname),  collection(collname), batchSize(100))
+    find({ salary: { $gt: $salary}});
````
It is indeed a reserved operator I will clarify that and make sure the manual with all the operators is linked.

delvecja commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,269 @@
+| ----------------- | -------------- |
+| b_date | STRING, INTEGER |
+| b_regex |  Unsupported |
+| b_timestamp | Unsupported |
```
It can be easily mapped to an integer, but not without losing some of the information. This is the format of the timestamp EJSON type: {"$timestamp":{"t":1565545664,"i":1}}. As you can see there are two fields that make up the value for timestamp and there isn't any clear documentation to what that "i" field is for. I didn't want to truncate it without understanding what the purpose of it is.

ghalliday replied:
I read somewhere it is a combination of a 32bit number of seconds since 1970 and a 32 bit sequence number.  I think it is fine to not support it for the moment.

delvecja commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,269 @@
+
+```
+dataset({STRING first, STRING last, REAL salary}) findMany(INTEGER include) := EMBED(mongodb : user(user), password(pwd), server(server), database(db),  collection(coll))
+    find({first: "John", last: "Doe"}, {first: $include, last: $include, salary: $include});
````
Yes, this is due to the fact that the plugin does not parse data types other than string. The argument for the MongoDB operation takes a 1 for true and 0 for false, so in order to have the plugin build a document with an integer value there needs to be a parameter passed in.

delvecja commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,695 @@
+         */
+        static MongoDBConnection& instance() 
+        {
+            static MongoDBConnection instance;
```
This is one of the nuances of MongoDB. It is very explicit in that it only allows a single instance object to be alive at one time. The only place I have been able to find where you can pass in connection options is in the instance constructor. The issue is that there is zero documentation on this and I have only been able to find a very simple example of these objects being used from MongoDB. I also feel as though there should be a way for each connection to have their own options even if they share an instance object I just haven't been able to find any information about it. This is something that will require more research and I will look into it more in depth.

ghalliday replied:
My reading of your code was that the configuration was supplied to the pool rather than the mongodb instance. So can you have multiple different pool instances that share the MongoDBConnection?  Beyond the scope of this PR, but please can you create a jira and add the number here so it can be revisited later.

delvecja replied:
That is my thought as well, but when I was starting the plugin I couldn't find any information about how that works with the pool objects. I have created a jira issue here: https://track.hpccsystems.com/browse/HPCC-28326

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+     * @param value Either the key for getting the value from view or a string to be inserted
+     */
+    template<typename T>
+    void insertPair(T &builder, const bsoncxx::document::view &view, std::string &key, bool isRsvd, std::string value)
```
Fixed

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+    {
+        Owned<MongoDBRowStream> mdbRowStream;
+        mdbRowStream.setown(new MongoDBRowStream(_resultAllocator, query));
+        return (byte *)mdbRowStream.getLink();
```
Yes that was incorrect, but I believe I still need to cast it as (byte *).

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+    void convertEJSONTypes(std::string &result, const char * &start, const char * &row, const char * &lastBrkt, int &depth) 
+    {
+        while (*start && *start != '$') start++;
+        auto end = start;
```
Fixed

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+            else
+                failx("Error processing result row");
+        }
+        return result;
```
I believe that was an artifact from the couchbase plugin that is no longer in use. It is removed and now returns nullptr.

delvecja commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,695 @@
+        std::string embeddedScript;                    //! Local copy of the embedded script.
+        std::string queryCMD;                          //! Local copy of the name of the MongoDB query operation.
+        bsoncxx::builder::basic::document builder{};   //! Local copy of the basic builder used for bind ECL rows to documents.
+        const char* cursor;                            //! Pointer for keeping track of parsing the embedded script.
```
Added default initializer. 
const char* cursor = nullptr;

delvecja commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,695 @@
+
+            queryCMD = std::string(script, end - script); // Get a string of the query command
+
+            embeddedScript = end + 1; // Save embedded script as what is after the parenthesis
```
Added if(embeddedScript.find("(") == std::string::npos) failx("Syntax Error: missing opening parenthesis");

ghalliday replied:
That isn't the correct test - embeddedScript hasn't been initialised at the point.  I think you want a check that *end != '('

delvecja replied:
Added check to ensure end lands on the opening parenthesis.

ghalliday commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2366 @@
+        if (script && *script) 
+        {
+            // Incoming script is not necessarily null terminated. Note that the chars refers to utf8 characters and not bytes.
+            size32_t len = rtlUtf8Size(chars, script);
```
minor: size is a better name than len (len is normally used for the number of characters/codepoints)

delvecja replied:
Changed

ghalliday commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,696 @@
+         * @param chars Size of the script.
+         * @param eScript pointer to beginning of the script.
+         */
+        void setEmbed(size32_t &chars, const char *eScript) 
```
The chars parameter is confusing.  It isn't really needed since eScript is null terminated in the calling code (since it is assigned to a StringAttr).  If you want to avoid that assignment then you should pass in the size and the original pointer, but the size does not need to be a reference parameter.

delvecja replied:
I just removed the parameter since the script is created using StringAttr making it redundant.

ghalliday commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,696 @@
+
+            if(embeddedScript.find("(") == std::string::npos) failx("Syntax Error: missing opening parenthesis");
+            end = script;
+            while(*end && (end - script) <= chars && *end != '(') end++; // Get pointer to end of the command
```
Confusion over whether the parameter is a null terminated string, or a string with a length.

delvecja replied:
Made the condition clearer as it is a null terminated string.

ghalliday commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,695 @@
+        unsigned int childrenProcessed;
+
+        // Simple constructor
+        PathTracker()
```
missed this comment

delvecja replied:
Removed because it was unused.

ghalliday commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+        rtlDataAttr data;
+        rtlStrToDataX(bytes, data.refdata(), len, value);
+
+        query->build()->append(kvp(std::string(field->name), std::string(data.getstr())));
```
Missed this comment

delvecja replied:
Apologies, it seems I jumped the gun. Value is a void pointer and cannot be passed into the string constructor. I have reverted back to the rtlStrToDataX call and added bytes to the string construction.

ghalliday reviewed (changes requested):
I may have followed up too quickly - because I realised you haven't addressed all the comments.  A few comments on your changes so far.

delvecja commented:
@ghalliday I appreciate your comments I should have things tested and ready by later today. Sorry, I tried to get most of your comments covered off yesterday due to my busy schedule and in my haste I wasn't able to fully test things. You're corrections are very much appreciated it would have taken me a little longer to realize where my mistakes were.

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+    {
+        checkNextParam(name);
+
+        query->build()->append(kvp(std::string(name), bsoncxx::types::b_utf8{std::string(val)}));
```
Fixed encoding. rtlDataAttr doesn't have a get() function, so where a char * was needed I used getstr() and for rtlUtf8Size it needed a void pointer and I used getdata().

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+    {
+        checkNextParam(name);
+
+        query->build()->append(kvp(std::string(name), bsoncxx::types::b_utf8{std::string(val)}));
```
Changed

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+    {
+        checkNextParam(name);
+
+        query->build()->append(kvp(std::string(name), bsoncxx::types::b_utf8{std::string(val)}));
```
Changed

delvecja commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2322 @@
+        char *utf8;
+        rtlUnicodeToUtf8X(utf8chars, utf8, chars, val);
+
+        query->build()->append(kvp(std::string(name), bsoncxx::types::b_utf8{std::string(utf8, utf8chars)}));
```
Changed

delvecja commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,695 @@
+         * @param pool The pool object that has the information about which database
+         * and collection to connect to.
+         */
+        void configure(std::unique_ptr<mongocxx::instance> instance, std::unique_ptr<mongocxx::pool> pool) 
```
Changed

delvecja commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,695 @@
+        const RtlTypeInfo *typeInfo;
+        const IContextLogger &logctx;
+        std::shared_ptr<MongoDBQuery> query;
+        int firstParam;
```
Added where I could. I don't think there is a default constructor for these members so for the ones that I could I initialized them as nullptr.

delvecja commented:
@ghalliday I believe I have responded to all of your comments and pushed the changes. Even though I opened a jira ticket I will look into the pool and see if there is a way to hand the connection options to that instead of the single instance object. I would like it if it could be part of this change, but there is simply not much documentation. I will see what I can do.

ghalliday commented:
Thanks I'll review tomorrow.  It is better to get this merged and then revisit the instance question than to delay trying to solve everything in one go.

ghalliday reviewed (approved):
@richardkchapman I think this is ready for merging.  Only thing is waiting for @AttilaVamos  to respond to the question about tags for the regression suite.  Also needs squashing.

AttilaVamos commented on testing/regress/ecl/mongodb-simple.ecl:
```diff
@@ -0,0 +1,92 @@
+Import MongoDB;
+
```
I think all of them would be useful.  

The first two to exclude it with all other plugins and embedded systems where no 3rd party server and/or library installed (included MongoDB) like Smoketest. 

The third one can be used to exclude that specific test or tests in an environment where all other 3rd party server and/or libraries are already installed. 

So, please add all of them to the this test code.

AttilaVamos commented on testing/regress/ecl/mongodb-simple.ecl:
```diff
@@ -0,0 +1,92 @@
+Import MongoDB;
```
Please add the usual Copyright header at the top of this file..

AttilaVamos reviewed (changes requested):
2 comments/requests.

delvecja commented:
@AttilaVamos Alright I added those flags and the copyright header. If everything is good I will go ahead and squash it down to one commit when you all are ready.

richardkchapman commented:
Looks good - can you squash to a single commit then I will merge

delvecja commented:
@richardkchapman Squashed it down to one commit.

AttilaVamos reviewed (approved):
Looks good
